cpp/src/traversal/opg/bfs_comms.cuh

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <raft/handle.hpp>
+#include <graph.hpp>
+
+
+namespace detail {
+
+namespace opg {
+
+constexpr inline size_t
+divup(const size_t& numerator, const size_t& denominator) {
+  return (numerator + denominator - 1)/denominator;
+}
+
+size_t number_of_words(size_t number_of_bits)
+{
+  size_t numerator = number_of_bits;
+  size_t denominator = BitsPWrd<unsigned>;
+  return (numerator > denominator) ?
+    1 + divup(numerator - denominator, denominator) :
+    (denominator > 0);
+}
+
+__global__
+void reduce_bitwise_or(
+    unsigned** frontiers,
+    size_t frontier_count,
+    int my_rank,
+    size_t word_count) {
+
+  unsigned word = 0;
+  size_t tid = threadIdx.x + (blockIdx.x * blockDim.x);
+  if (tid < word_count) {
+    for (int i = 0; i < frontier_count; ++i) {
```

Comment:
If I am not mistaken, frontier_count here is the # of GPUs. So, the amount of computing here actually increases with an increasing number of GPUs. This will kill scalability unless you are expecting this code to work only for a small number of GPUs.

Reply:
This is a standin code for nccl bcast with bitwise or. Having said that, the computation required for an allreduce operation will increase with an increasing number of GPUs.